test/shared/Governance.ts

```diff
@@ -0,0 +1,63 @@
+import {Signer} from 'ethers';
+import {ethers} from 'hardhat';
+import {impersonateAccountWithFunds, stopImpersonateAccount} from '../shared/AccountManipulation';
+import {increaseNextBlockTime, setNextBlockNumber} from '../shared/TimeManipulation';
+import {DeployedContracts} from './Forking';
+
+export const createAndExecuteProposal = async ({
+  governor,
+  timeLock,
+  arenaToken,
+  user,
+  targets,
+  values,
+  calldatas,
+}: {
+  user: Signer;
+  targets: string[];
+  values: string[];
+  calldatas: string[];
+} & DeployedContracts) => {
+  // we need address with min. proposalThreshold tokens to propose
+  // 1. borrow some treasury tokens to user
+  const quorumAmount = await governor.quorumVotes();
+  // careful, this sends ETH to timelock which might break real-world simulation for proposals involving Timelock ETH
+  const timeLockSigner = await impersonateAccountWithFunds(timeLock.address);
+  await arenaToken.connect(timeLockSigner).transfer(await user.getAddress(), quorumAmount);
+  // set voting delay & duration to 2 blocks, otherwise need to simulate 302,400 blocks
+  await governor.connect(timeLockSigner).setVotingDelay(`2`);
+  await governor.connect(timeLockSigner).setVotingPeriod(`2`);
+  await stopImpersonateAccount(timeLock.address);
+
+  await arenaToken.connect(user).delegate(await user.getAddress());
+  const descriptionHash = ethers.utils.keccak256([]); // keccak(``)
+  let tx = await governor.connect(user)['propose(address[],uint256[],bytes[],string)'](targets, values, calldatas, ``);
+  let {events} = await tx.wait();
+  // get first event (ProposalCreated), then get first arg of that event (proposalId)
+  const proposalId: string = events![0].args![0].toString();
+
+  // 2. advance time past voting delay and vote on proposal
+  const voteStartBlock = await governor.proposalSnapshot(proposalId);
+  await setNextBlockNumber(voteStartBlock.toNumber() + 1); // is a blocknumber which fits in Number
+  tx = await governor.connect(user)['castVote'](proposalId, `1`);
+
+  // 3. return borrowed tokens
+  tx = await arenaToken.connect(user).transfer(timeLock.address, quorumAmount);
+
+  // 4. advance time past voting period and queue proposal calls to Timelock via GovernorTimelockControl.queue
+  const voteEndBlock = await governor.proposalDeadline(proposalId);
+  await setNextBlockNumber(voteEndBlock.toNumber() + 1); // is a blocknumber which fits in Number
```

Comment:
would prefer to call `setNextBlockTimestamp` here to advance time by 1 week. This is to better simulate time-sensitive operations. For instance, our token sale is only set to commence from a certain date.

Reply:
I need to mine blocks here because the Governor uses blocks to measure time (interestingly, the Timelock uses seconds) but I can in addition increase the time the blocks are minted at

Reply:
Changed, please have a look again